src/units/animation.cpp

```diff
@@ -1329,6 +1373,196 @@ void unit_animator::add_animation(unit_const_ptr animated_unit
+		auto& anim = animated_units_[temp_index];
+		// either create a new vector when fail to match existing unit, or append to last matched unit vector.
+		bool new_queue = true;
+		for(auto iter = processed_anim_queues.begin(); iter != processed_anim_queues.cend(); iter++) {
+			auto& sorted_anim = (*iter)[0].get();
+			if(sorted_anim.move_up == anim.move_up) {
+				(*iter).emplace_back(std::ref(anim));
+				new_queue = false;
+				break;
+			}
+		}
+		if(new_queue) {
+			processed_anim_queues.emplace_back(std::vector<std::reference_wrapper<anim_elem>>{std::ref(anim)});
+		}
+		temp_index++;
+	}
+	anim_elem* last_anim_p = nullptr;
+	std::vector<std::vector<config>> uconfig_queues;
+	std::vector<std::pair<std::reference_wrapper<unit_ptr>, map_location>> initial_uloc_pairs;
+	// generate configs judged if the moving can be coherent
+	for(auto& vec : processed_anim_queues) {
+		std::vector<config> judged_uconfigs;
+		for(auto iter = vec.begin(); iter != vec.cend(); iter++) {
+			auto& anim = (*iter).get();
+			coherent_moving
+				= last_anim_p && anim.coherence && last_anim_p->move_up == anim.move_up && last_anim_p->src == anim.src;
+			unit_ptr& m_up = anim.move_up;
+			unit& m_u = *m_up;
+			map_location& m_src = coherent_moving ? last_anim_p->move_dst : anim.src;
+			map_location& m_dst = anim.move_dst;
+
+			bool pair_found = false;
+			for (auto& elem : initial_uloc_pairs) {
+				if (elem.first.get() == m_up) {
+					pair_found = true;
+					break;
+				}
+			}
+			if (!pair_found) {
+				initial_uloc_pairs.emplace_back(std::ref(m_up), m_u.get_location());
+			}
+
+			m_u.set_hidden(true);
+			if(coherent_moving) {
+				m_u.set_location(last_anim_p->move_dst);
+			} else {
+				m_u.set_location(map_location::null_location());
+			}
+			disp->invalidate(m_u.get_location());
+			m_dst = find_vacant_tile(m_dst, pathfind::VACANT_ANY, m_up.get());
+			config cfg;
+			cfg["type"] = m_u.type().id();
+			cfg["gender"] = m_u.gender();
+			cfg["variation"] = m_u.variation();
+			cfg["side"] = m_u.side();
+			cfg["x"] = std::to_string(m_src.wml_x()) + "," + std::to_string(m_dst.wml_x());
+			cfg["y"] = std::to_string(m_src.wml_y()) + "," + std::to_string(m_dst.wml_y());
+			judged_uconfigs.emplace_back(cfg);
+			last_anim_p = &anim;
+		}
+		uconfig_queues.emplace_back(judged_uconfigs);
+	}
+	// reset locations so that when the animated units are reverted from hidden, they will show up on screen normally.
+	for (auto iter = initial_uloc_pairs.begin(); iter != initial_uloc_pairs.cend(); iter++) {
+		auto& uloc = *(iter);
+		unit& m_u = *(uloc.first.get());
+		m_u.set_location(uloc.second);
+		m_u.anim_comp().set_standing();
+	}
+	// generate the needed fake_units by extracting one from each unit's config sequence
+	while(!uconfig_queues.empty()) {
+		config cfg;
+		for(auto index = 0; index < uconfig_queues.size(); index++) {
+			auto& vec = uconfig_queues[index];
+			if(vec.empty()) {
+				uconfig_queues.erase(uconfig_queues.begin() + index);
```

Comment:
`uconfig_queues` is local to this function, right? In that case, the additional bookkeeping (erasing) isn't necessary, since the list is going to be destroyed once the function exits anyway. All you need is the config data to invoke the desired tag.

But see my comment above regarding possibly doing away with the extra lists.